iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c

```diff
@@ -0,0 +1,441 @@
+        result = g_statusSuccess;
+    }
+
+    json_value_free(rootValue);
+    free(jsonStr);
+
+    return result;
+}
+
+//
+// GetDesiredJson retrieves JSON_Object* in the JSON tree corresponding to the desired payload.
+//
+static JSON_Object* GetDesiredJson(DEVICE_TWIN_UPDATE_STATE updateState, JSON_Value* rootValue)
+{
+    JSON_Object* rootObject = NULL;
+    JSON_Object* desiredObject;
+
+    if ((rootObject = json_value_get_object(rootValue)) == NULL)
+    {
+        LogError("Unable to get root object of JSON");
+        desiredObject = NULL;
+    }
+    else
+    {
+        if (updateState == DEVICE_TWIN_UPDATE_COMPLETE)
+        {
+            // For a complete update, the JSON from IoTHub will contain both "desired" and "reported" - the full twin.
+            // We only care about "desired" in this sample, so just retrieve it.
+            desiredObject = json_object_get_object(rootObject, "desired");
+        }
+        else
+        {
+            // For a patch update, IoTHub only sends "desired" portion of twin and does not explicitly put a "desired:" JSON envelope.
+            // So here we simply need the root of the JSON itself.
+            desiredObject = rootObject;
+        }
+    }
+
+    return desiredObject;
+}
+
+//
+// UpdateTemperatureAndStatistics updates the temperature and min/max/average values
+//
+static void UpdateTemperatureAndStatistics(double desiredTemp, bool* maxTempUpdated)
+{
+    if (desiredTemp > g_maxTemperature)
+    {
+        g_maxTemperature = desiredTemp;
+        *maxTempUpdated = true;
+    }
+    else if (desiredTemp < g_minTemperature)
+    {
+        g_minTemperature = desiredTemp;
+    }
+
+    g_numTemperatureUpdates++;
+    g_allTemperatures += desiredTemp;
+
+    g_currentTemperature = desiredTemp;
+}
+
+// Format string to report the property maximum temperature since reboot.  This is a "read only" property from the
+// service solution's perspective, which means we don't need to include any sort of status codes.
+static const char g_maxTemperatureSinceRebootFormat[] = "{\"maxTempSinceLastReboot\": %.2f }";
+
+// Format string to indicate the device received an update request for the temperature.  Because this is a "writeable"
+// property from the service solution's perspective, we need to return a status code (HTTP status code style) and version
+// for the solution to correlate the request and its status.
+static const char g_targetTemperatureResponseFormat[] = "{\"targetTemperature\": { \"value\": %.2f, \"ac\":%d, \"av\":%d, \"ad\":\"%s\" }}";
+
+
+//
+// SendTargetTemperatureReport sends a PnP property indicating the device has received the desired targetted temperature
+//
+static void SendTargetTemperatureReport(IOTHUB_DEVICE_CLIENT_HANDLE deviceClient, double desiredTemp, int responseStatus, int version, const char* description)
+{
+    IOTHUB_CLIENT_RESULT iothubClientResult;
+    char targetTemperatureResponseProperty[256];
```

Comment:
nit: large buffer size on the stack. consider making it static at the top

Reply:
This is a hard one....  but I'm going to leave as-is.

My gut though is that current C SDK can take a stack hit this size (it'd be different on embedded) and I'd feel safer with that rather than making this a global.  I get way we work today it'd be OK, but at some point we could imagine `SendTargetTemperatureReport` being called from the `main` in which case there could be multiple threads operating on the buffer.